sale_import_base/models/sale_order.py

```diff
@@ -0,0 +1,370 @@
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+from copy import deepcopy
+from odoo.addons.queue_job.job import job
+
+MAPPINGS_SALE_ORDER_ADDRESS_SIMPLE = [
+    "name",
+    "street",
+    "street2",
+    "zip",
+    "city",
+    "email",
+]
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    si_exc_check_amounts_untaxed = fields.Boolean(
+        "(technical) Check untaxed amounts against imported values"
+    )
+    si_exc_check_amounts_total = fields.Boolean(
+        "(technical) Check total amounts against imported values"
+    )
+    si_amount_untaxed = fields.Float("(technical) Untaxed amount from import")
+    si_amount_tax = fields.Float("(technical) Tax amount from import")
+    si_amount_total = fields.Float("(technical) Total amount from import")
+
+    def batch_process_json_imports(self, imports, **kwargs):
+        """
+        For an up-to-date version of the imports format, check:
+            1. datamodels
+            2. _si_validate_datamodel function
+        kwargs effectively function as way to pass context to jobs
+        methods prefixed by "si" are related to sale imports
+        """
+        jobs = self.env["queue.job"]
+        for el in imports:
+            new_job = self.with_delay(self.process_json_import(el, kwargs))
+            jobs += new_job
+        return jobs
+
+    @job
+    def process_json_import(self, data, **kwargs):
+        errors = self.env.datamodels["sale.order"].validate(data)
+        if errors:
+            raise ValidationError(_("Import could not be validated: %s" % errors))
+        so_datamodel = self.env.datamodels["sale.order"].load(data)
+        # validation and value processing: we could extend datamodel with the same idea
+        self._si_validate_datamodel(so_datamodel)
+        so_dump_raw = so_datamodel.dump()
+        so_vals = self._si_process_dump(so_dump_raw)
+        new_sale_order = self.env["sale.order"].create(so_vals)
+        new_sale_order._si_finalize(new_sale_order, data)
+        return new_sale_order
+
+    # DATAMODEL VALIDATORS
+    def _si_validate_datamodel(self, datamodel_instance):
+        """ Extend this method to add your validators against the DB """
+        # self._si_validate_address_customer(datamodel_instance)
+        self._si_validate_product_codes(datamodel_instance)
+        self._si_validate_sale_channel(datamodel_instance)
+        self._si_validate_currency_code(datamodel_instance)
+        self._si_validate_payment(datamodel_instance)
+
+    # def _si_validate_address_customer(self, datamodel_instance):
+    #     partner = self.env["res.partner"].search(
+    #         [("email", "=", datamodel_instance.address_customer.email)]
+    #     )
+    #     if len(partner.ids) != 1:
+    #         raise ValidationError(_("Could not find one partner"))
+
+    def _si_validate_product_codes(self, datamodel_instance):
```

Comment:
If I am not wrong this logic can be moved inside the validator as we have the env now ? right?